lab3van/util-linux/disk-utils/swaplabel.8
.\" Copyright 2010 Jason Borden <jborden@bluehost.com>
.\"
.\" This file may be copied under the terms of the GNU Public License.
.\"
.TH SWAPLABEL 8 "April 2010" "util-linux" "System Administration"
.SH NAME
swaplabel \- print or change the label or UUID of a swap area
.SH SYNOPSIS
.B swaplabel
.RB [ \-L
.IR label ]
.RB [ \-U
.IR UUID ]
.I device
.SH DESCRIPTION
.B swaplabel
will display or change the label or UUID of a swap partition located on
<<<<<<< HEAD
.IR device 
(or regular file).
.PP
If the optional arguments
.B \-L 
and
.B \-U 
=======
.I device
(or regular file).
.PP
If the optional arguments
.B \-L
and
.B \-U
>>>>>>> master-vanilla
are not given,
.B swaplabel
will simply display the current swap-area label and UUID of
.IR device .
.PP
If an optional argument is present, then
.B swaplabel
will change the appropriate value on
.IR device .
These values can also be set during swap creation using
.BR mkswap (8).
The
.B swaplabel
utility allows to change the label or UUID on an actively used swap device.
.SH OPTIONS
.TP
.BR \-h , " \-\-help"
Display help text and exit.
.TP
.BR \-L , " \-\-label " \fIlabel\fR
Specify a new \fIlabel\fR for the device.
Swap partition labels can be at most 16 characters long.  If
.I label
is longer than 16 characters,
.B swaplabel
will truncate it and print a warning message.
.TP
.BR \-U , " \-\-uuid " \fIUUID\fR
Specify a new \fIUUID\fR for the device.
The \fI UUID\fR
must be in the standard 8-4-4-4-12 character format, such as is output by
.BR uuidgen (1) .
.PP
.SH AUTHOR
.B swaplabel
was written by Jason Borden <jborden@bluehost.com> and Karel Zak <kzak@redhat.com>.
<<<<<<< HEAD
=======
.SH ENVIRONMENT
.IP LIBBLKID_DEBUG=all
enables libblkid debug output.
>>>>>>> master-vanilla
.SH AVAILABILITY
The swaplabel command is part of the util-linux package and is available from
ftp://ftp.kernel.org/pub/linux/utils/util-linux/.
.SH SEE ALSO
.BR mkswap (8),
.BR swapon (8),
.BR uuidgen (1)


debian/changelog
util-linux (2.27.160-1) unstable; urgency=low

  * Initial release (Closes: #nnnn)  <nnnn is the bug number of your ITP>

 -- roma <newangel88@mail.ru>  Thu, 24 Dec 2015 13:07:17 +0400


lab3van/util-linux/disk-utils/swaplabel.c
/*
 * swaplabel.c - Print or change the label / UUID of a swap partition
 *
 * Copyright (C) 2010 Jason Borden <jborden@bluehost.com>
 * Copyright (C) 2010 Karel Zak <kzak@redhat.com>
 *
 * Usage: swaplabel [-L label] [-U UUID] device
 *
 * This file may be redistributed under the terms of the GNU Public License
 * version 2 or later.
 *
 */
#include <stdio.h>
#include <stddef.h>
#include <string.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdlib.h>
<<<<<<< HEAD
#include <blkid.h>
=======
>>>>>>> master-vanilla
#include <getopt.h>

#ifdef HAVE_LIBUUID
# include <uuid.h>
#endif

#include "c.h"
<<<<<<< HEAD
#include "writeall.h"
#include "swapheader.h"
#include "strutils.h"
#include "nls.h"
=======
#include "nls.h"
#include "all-io.h"
#include "strutils.h"
#include "closestream.h"

#include "swapheader.h"
#include "swapprober.h"
>>>>>>> master-vanilla

#define SWAP_UUID_OFFSET	(offsetof(struct swap_header_v1_2, uuid))
#define SWAP_LABEL_OFFSET	(offsetof(struct swap_header_v1_2, volume_name))

<<<<<<< HEAD
/*
 * Returns new libblkid prober. This function call exit() on error.
 */
static blkid_probe get_swap_prober(const char *devname)
{
	blkid_probe pr;
	int rc;
	const char *version = NULL;
	char *swap_filter[] = { "swap", NULL };

	pr = blkid_new_probe_from_filename(devname);
	if (!pr) {
		warn(_("%s: unable to probe device"), devname);
		return NULL;
	}

	blkid_probe_enable_superblocks(pr, TRUE);
	blkid_probe_set_superblocks_flags(pr,
			BLKID_SUBLKS_LABEL | BLKID_SUBLKS_UUID |
			BLKID_SUBLKS_VERSION);

	blkid_probe_filter_superblocks_type(pr, BLKID_FLTR_ONLYIN, swap_filter);

	rc = blkid_do_safeprobe(pr);
	if (rc == -1)
		warn(_("%s: unable to probe device"), devname);
	else if (rc == -2)
		warnx(_("%s: ambivalent probing result, use wipefs(8)"), devname);
	else if (rc == 1)
		warnx(_("%s: not a valid swap partition"), devname);

	if (rc == 0) {
		/* supported is SWAPSPACE2 only */
		blkid_probe_lookup_value(pr, "VERSION", &version, NULL);
		if (strcmp(version, "2"))
			warnx(_("%s: unsupported swap version '%s'"),
						devname, version);
		else
			return pr;
	}

	blkid_free_probe(pr);
	return NULL;
}

=======
>>>>>>> master-vanilla
/* Print the swap partition information */
static int print_info(blkid_probe pr)
{
	const char *data;

	if (!blkid_probe_lookup_value(pr, "LABEL", &data, NULL))
		printf("LABEL: %s\n", data);

	if (!blkid_probe_lookup_value(pr, "UUID", &data, NULL))
		printf("UUID:  %s\n", data);

	return 0;
}

/* Change the swap partition info */
static int change_info(const char *devname, const char *label, const char *uuid)
{
	int fd;

	fd = open(devname, O_RDWR);
	if (fd < 0) {
<<<<<<< HEAD
		warn(_("%s: failed to open"), devname);
=======
		warn(_("cannot open %s"), devname);
>>>>>>> master-vanilla
		goto err;
	}
#ifdef HAVE_LIBUUID
	/* Write the uuid if it was provided */
	if (uuid) {
		uuid_t newuuid;

		if (uuid_parse(uuid, newuuid) == -1)
			warnx(_("failed to parse UUID: %s"), uuid);
		else {
			if (lseek(fd, SWAP_UUID_OFFSET, SEEK_SET) !=
							SWAP_UUID_OFFSET) {
				warn(_("%s: failed to seek to swap UUID"), devname);
				goto err;

			} else if (write_all(fd, newuuid, sizeof(newuuid))) {
				warn(_("%s: failed to write UUID"), devname);
				goto err;
			}
		}
	}
#endif
	/* Write the label if it was provided */
	if (label) {
		char newlabel[SWAP_LABEL_LENGTH];

		if (lseek(fd, SWAP_LABEL_OFFSET, SEEK_SET) != SWAP_LABEL_OFFSET) {
			warn(_("%s: failed to seek to swap label "), devname);
			goto err;
		}
		memset(newlabel, 0, sizeof(newlabel));
		xstrncpy(newlabel, label, sizeof(newlabel));

		if (strlen(label) > strlen(newlabel))
			warnx(_("label is too long. Truncating it to '%s'"),
					newlabel);
		if (write_all(fd, newlabel, sizeof(newlabel))) {
			warn(_("%s: failed to write label"), devname);
			goto err;
		}
	}

<<<<<<< HEAD
	close(fd);
=======
	if (close_fd(fd) != 0) {
		warn(_("write failed: %s"), devname);
		return -1;
	}
>>>>>>> master-vanilla
	return 0;
err:
	if (fd >= 0)
		close(fd);
	return -1;
}

static void __attribute__((__noreturn__)) usage(FILE *out)
{
<<<<<<< HEAD
	fprintf(out, _("Usage: %s [options] <device>\n\nOptions:\n"),
			program_invocation_short_name);

	fprintf(out, _(
		" -h, --help          this help\n"
		" -L, --label <label> specify a new label\n"
		" -U, --uuid <uuid>   specify a new uuid\n"));

	fprintf(out, _("\nFor more information see swaplabel(8).\n"));

=======
	fputs(USAGE_HEADER, out);
	fprintf(out, _(" %s [options] <device>\n"),
		program_invocation_short_name);

	fputs(USAGE_SEPARATOR, out);
	fputs(_("Display or change the label or UUID of a swap area.\n"), out);

	fputs(USAGE_OPTIONS, out);
	fputs(_(" -L, --label <label> specify a new label\n"
		" -U, --uuid <uuid>   specify a new uuid\n"), out);
	fputs(USAGE_SEPARATOR, out);
	fputs(USAGE_HELP, out);
	fputs(USAGE_VERSION, out);
	fprintf(out, USAGE_MAN_TAIL("swaplabel(8)"));
>>>>>>> master-vanilla
	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

int main(int argc, char *argv[])
{
	blkid_probe pr = NULL;
	char *uuid = NULL, *label = NULL, *devname;
	int c, rc = -1;

	static const struct option longopts[] = {
	    { "help",      0, 0, 'h' },
<<<<<<< HEAD
=======
	    { "version",   0, 0, 'V' },
>>>>>>> master-vanilla
	    { "label",     1, 0, 'L' },
	    { "uuid",      1, 0, 'U' },
	    { NULL,        0, 0, 0 }
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
<<<<<<< HEAD

	while ((c = getopt_long(argc, argv, "hL:U:", longopts, NULL)) != -1) {
=======
	atexit(close_stdout);

	while ((c = getopt_long(argc, argv, "hVL:U:", longopts, NULL)) != -1) {
>>>>>>> master-vanilla
		switch (c) {
		case 'h':
			usage(stdout);
			break;
<<<<<<< HEAD
=======
		case 'V':
			printf(UTIL_LINUX_VERSION);
			return EXIT_SUCCESS;
>>>>>>> master-vanilla
		case 'L':
			label = optarg;
			break;
		case 'U':
#ifdef HAVE_LIBUUID
			uuid = optarg;
#else
			warnx(_("ignore -U (UUIDs are unsupported)"));
#endif
			break;
		default:
			usage(stderr);
			break;
		}
	}

	if (optind == argc)
		usage(stderr);

	devname = argv[optind];
	pr = get_swap_prober(devname);
	if (pr) {
		if (uuid || label)
			rc = change_info(devname, label, uuid);
		else
			rc  = print_info(pr);
		blkid_free_probe(pr);
	}
	return rc ? EXIT_FAILURE : EXIT_SUCCESS;
}

